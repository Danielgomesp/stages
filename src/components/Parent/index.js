import React, { useState } from 'react';
import { Container, Heading, Paragraph } from 'styled-minimal';
import { steps, durations } from 'data.json';

import { mapSteps } from 'helpers';

import Child from '../Child';

import 'styles.css';

// O mappedSteps recebe um array com todos os dados que eu preciso.
const mappedSteps = mapSteps(steps, durations);

// A ideia é: A cada iteração, eu removo um item do mappedSteps e adiciono ao estado fila(queue).
// A partir desse estado eu faço o looping renderizando o Child.
// Começando aqui, será removido o 1º item do mappedSteps e passo como primeiro estado da fila.
const initialState = [mappedSteps.shift()];

export default function Parent() {
  const [queue, setQueue] = useState(initialState);
  const [totalDuration, setTotalDuration] = useState(0);
  const [haveFinished, setHaveFinished] = useState(false);

  function handleCallBack() {
    const currentItem = queue[queue.length - 1]; //  Recebe último item da fila.

    // Aqui está a 'mágica'.
    // Aguardo o tempo (duration) de cada item para então
    // atualizar o estado.
    setTimeout(() => {
      setTotalDuration(totalDuration + currentItem.duration);

      if (mappedSteps.length === 0) {
        setHaveFinished(true);
        return;
      }

      setQueue([
        ...queue,
        mappedSteps.shift(),
      ]);
    }, currentItem.duration);
  }

  return (
    <Container data-testid="ParentElement">
      <Heading mb={4} textAlign="center">
        Stages
      </Heading>
      {
        queue.map((item) => (
          <Child
            key={item.key}
            color={item.value}
            callBackHandler={handleCallBack}
            duration={item.duration}
          />
        ))
      }
      <div>
        {
          haveFinished
          // hehehe
          && (
            <>
              <Paragraph fontWeight="600" marginTop="30px">
                {`Tempo total: ${totalDuration}ms`}
              </Paragraph>
              <img
                alt="OMG, we should hire him"
                src="https://blogpilates.com.br/wp-content/uploads/2019/06/gif-uau.gif"
                width="280"
              />
            </>
          )
        }
      </div>
    </Container>
  );
}
