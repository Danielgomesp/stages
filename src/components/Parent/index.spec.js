import React from 'react';
import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Parent from './index';

describe('Parent', () => {
  it('Should render Parent and one Child on the screen', () => {
    const { container } = render(<Parent />);
    expect(container).toMatchSnapshot();
  });

  it('Should render all 4 Childs on the screen', () => {
    jest.useFakeTimers();

    render(<Parent />);

    act(() => {
      jest.runAllTimers(); // trigger setTimeout
    });
    expect(screen.getAllByTestId('ChildElement')).toHaveLength(2);

    act(() => {
      jest.runAllTimers(); // trigger setTimeout
    });
    expect(screen.getAllByTestId('ChildElement')).toHaveLength(3);

    act(() => {
      jest.runAllTimers(); // trigger setTimeout
    });
    expect(screen.getAllByTestId('ChildElement')).toHaveLength(4);

    act(() => {
      jest.runAllTimers();
    });

  });

  it('Should render a final message on Parent', () => {
    render(<Parent />);

    act(() => {
      jest.runAllTimers(); // trigger setTimeout
    });

    expect(screen.getByAltText('OMG, we should hire him')).toBeDefined();
  })
});
