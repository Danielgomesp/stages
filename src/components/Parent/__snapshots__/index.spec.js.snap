// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parent Should render Parent and one Child on the screen 1`] = `
<div>
  <div
    class="sc-fzoLsD CSqIi"
    data-testid="ParentElement"
  >
    <h1
      class="sc-fzpjYC bGlwZc"
    >
      Stages
    </h1>
    <div
      class="sc-AxjAm jUgobI"
      data-testid="ChildElement"
      display="flex"
      font-weight="700"
      height="128"
    >
      <div
        class="sc-fzoXWK fcyArp"
        color="#FFEB3B"
        opacity="0.7"
        size="48"
        speed="2"
      >
        <div />
        <div />
      </div>
    </div>
    <div />
  </div>
</div>
`;
